designs/carbon-aware.md

```diff
@@ -0,0 +1,197 @@
+
+#### Considerations
+1. 👍 Current consolidation methods are unaffected.
+1. 👎 There might be copy-paste of code from the original consolidation methods to the carbon efficient consolidators.
+
+### Option 2: Use Carbon Efficient filtering/sorting methods
+
+#### Consolidation
+Create carbon efficient implementations of low-level functions like `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that are used when Carbon Efficient is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.
+
+#### Provisioning
+Use same changes to provisioning as in [option 1](#option-1-use-carbon-efficient-provisioning-and-concolidation-methods).
+
+#### Considerations
+1. 👍 Less code copy-paste.
+1. 👍 Improvements to original consolidation methods also improve the Carbon Efficient feature.
+1. 👎 Has a risk of breaking undocumented invariants.
+1. 👎 Adds complexity to the original consolidation methods.
+
+### Option 3: Override instance price with carbon price (recommended)
+Minimize carbon emissions by defining a price per kgCO₂e and override the instance price with the carbon price (USD/kgCO₂e). Using the `prioritized` launch strategy, carbon emissions will be minimized during provisioning. Consolidation will unknowingly consolidate to minimize carbon emissions.
+
+The carbon price will depend on `region` and `instanceType` and assume constant resource utilization (e.g. always 80% utilization). The carbon price will be generated in a "hack" and included as consts (same method as used for generating initial pricing[^2]). The carbon price / emission estimates can be updated with new versions.
+
+Another feature (added later) can be to add carbon price to instance price to simulate a [carbon tax](https://en.wikipedia.org/wiki/Carbon_tax). Administrators could configure a custom carbon price or use a default.
+
+#### Considerations
+1. 👍 Change is constrained to the pricing domain, so most of Karpenter's logic remains unaffected.
+1. 👍👍 A simulated carbon tax could be appealing for *Beta* or *General Availability*[^3] as it combines the real price with the carbon price.
+1. 👎 Adds complexity to the *price* concept. Price is not *just* price, but rather becomes an optimization function.
+1. 👎 Depending on implementation, the `karpenter_cloudprovider_instance_type_price_estimate` metric *may* represent more than just price when Carbon Efficient is enabled.
+
+### Option 4: Enable custom instance price overrides
+Enable administrators to configure custom instance price overrides, e.g. in a ConfigMap. A configuration using emission factors (varying with region and instance type) masked as prices can be pre-generated. Administrators then copy-paste a Carbon Efficient `priceOverride` into their environment.
+
+```yaml
+priceOverrides:
+  - instanceType: "m5.large"
+    region: "eu-west-1"
```

Comment:
Price overrides is interested in multiple respects. For example there are reserved instances customers will reserve on azure and get a better contract for. So its actually cheaper for karpenter ot use those instances they got at a discount than the typical "retail pricing api" pricing data for the instance price. 


So this field of pricing overrides can be used and extended to other usecases.

Reply:
Yes, I have heard in multiple occasions that price overrides could be useful for various reasons. 

In my [Master's Thesis](https://github.com/JacobValdemar/carbon-efficient-karpenter-thesis), overriding the price of all existing instance types is actually what I did. It is feasible and maintainable with enough automation. I think it makes sense in _this_ context, but might be infeasible in other use cases.